src/bar.rs

```diff
@@ -81,74 +144,102 @@ impl Bar {
             });
 
         Box::new(future)
-    }
-
+    } */
     /// Redraws components in the center slot.
     /// Unforunately centering means that all components
     /// must be redrawn if even one of them changes size.
     fn redraw_center(&mut self) -> Result<()> {
         // Get future width of all center components
-        let width_all: u16 = self.center_items
+        let width_all: u16 = self.slot_items(Slot::Center)
             .iter()
-            .map(|item| item.get_content_width())
+            .map(|item| item.0.width().unwrap_or(0))
             .sum();
 
         // Draw blank background to prevent leftovers after shrinkage
         // Only does this when component width has shrunk
-        let old_width_all: u16 = self.center_items
+        let old_width_all: u16 = self.slot_items(Slot::Center)
             .iter()
-            .map(|item| self.item_positions[item.get_id()].1)
+            .filter_map(|&(ref ctx, _)| ctx.width())
             .sum();
+
         if width_all < old_width_all {
-            if let Some(first) = self.center_items.first() {
-                let old_start = self.item_positions[first.get_id()].0;
-                self.paint_bg(old_start, old_start + old_width_all)?;
+            if let Some(&(ref ctx, _)) = self.slot_items(Slot::Center).first() {
+                if let Some(old_start) = ctx.position {
+                    self.paint_bg(old_start, old_start + old_width_all)?;
+                }
```

Comment:
I also found the issue that when position changes, but not width, there will be no redraw. Getting the old position works tho. So this is an oversight in my initial implementation.